File: reactive-logger-example-java/src/test/java/io/github/numichi/reactive/logger/example/java/ExampleComponentTest.java
package io.github.numichi.reactive.logger.example.java;

import io.github.numichi.reactive.logger.Configuration;
import io.github.numichi.reactive.logger.exceptions.ReadException;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.mockito.Mockito;
import org.slf4j.Logger;
import reactor.test.StepVerifier;

import java.util.Map;

import static org.mockito.Mockito.*;

class ExampleComponentTest {
    
    private final Logger logger = Mockito.mock(Logger.class);
    private final ExampleComponent component = new ExampleComponent(logger);
    
    @BeforeEach
    void beforeEach() {
        Configuration.reset();
        Mockito.reset(logger);
    }
    
    @Test
    void example1Test() {
        StepVerifier.create(component.example1())
            .expectNext(Map.of("foo", "bar"))
            .verifyComplete();
        
        verify(logger).info("example1");
    }
    
    @Test
    void example2Test() {
        StepVerifier.create(component.example2()).verifyComplete();
        
        verify(logger, times(0)).trace(any());
        verify(logger, times(1)).debug("bar");
        verify(logger, times(1)).info("null");
        verify(logger, times(0)).warn(any());
    }
    
    @Test
    void example3Test() {
        StepVerifier.create(component.example3()).verifyComplete();
        
        verify(logger, times(0)).trace(any());
        verify(logger, times(1)).debug("bar");
        verify(logger, times(1)).info("null");
        verify(logger, times(0)).warn(any());
    }
    
    @Test
    void example4Test() {
        StepVerifier.create(component.example4()).verifyError(ReadException.class);

        verify(logger).error("DEFAULT_REACTOR_CONTEXT_MDC_KEY context key is not contain in context");
    }
}

File: reactive-logger-example-kotlin/src/test/kotlin/io/github/numichi/reactive/logger/example/kotlin/ExampleComponentTest.kt
package io.github.numichi.reactive.logger.example.kotlin

import io.github.numichi.reactive.logger.Configuration
import io.github.numichi.reactive.logger.exceptions.ReadException
import io.mockk.clearMocks
import io.mockk.mockk
import io.mockk.verify
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.test.runTest
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertThrows
import org.slf4j.Logger

@ExperimentalCoroutinesApi
internal class ExampleComponentTest {
    private val logger = mockk<Logger>(relaxed = true)
    private val component = ExampleComponent(logger)

    @BeforeEach
    fun beforeEach() {
        Configuration.reset()
        clearMocks(logger)
    }

    @Test
    fun example1Test() = runTest {
        val result = component.example1()

        assertEquals(mapOf("foo" to "bar"), result)

        verify(exactly = 1) { logger.info("example1") }
    }

    @Test
    fun example2Test() = runTest {
        component.example2()

        verify(exactly = 0) { logger.trace(any()) }
        verify(exactly = 2) { logger.debug("bar") }
        verify(exactly = 1) { logger.info("null") }
        verify(exactly = 0) { logger.warn(any()) }
    }

    @Test
    fun example3Test() = runTest {
        component.example3()

        verify(exactly = 0) { logger.trace(any()) }
        verify(exactly = 2) { logger.debug("bar") }
        verify(exactly = 1) { logger.info("null") }
        verify(exactly = 0) { logger.warn(any()) }
    }

    @Test
    fun example4Test() = runTest {
        val throwable = assertThrows<ReadException> {
            component.example4()
        }

        val expected = "DEFAULT_REACTOR_CONTEXT_MDC_KEY context key is not contain in context"

        assertEquals(expected, throwable.message)
        verify(exactly = 1) { logger.error(expected) }
    }
}